File: victorious/victorious/UI/Like/VLikersDataSource.m
//
//  VLikersDataSource.m
//  victorious
//
//  Created by Patrick Lynch on 6/17/15.
//  Copyright (c) 2015 Victorious. All rights reserved.
//

#import "VLikersDataSource.h"
#import "VObjectManager+Pagination.h"
#import "VSequence.h"

@interface VLikersDataSource ()

@property (nonatomic, strong) VSequence *sequence;

@end

@implementation VLikersDataSource

- (instancetype)initWithSequence:(VSequence *)sequence
{
    NSParameterAssert( sequence != nil );
    
    self = [super init];
    if ( self != nil )
    {
        _sequence = sequence;
    }
    return self;
}

#pragma mark - VUsersDataSource

- (NSString *)noContentTitle
{
    return NSLocalizedString( @"NoLikersTitle", @"" );
}

- (NSString *)noContentMessage
{
    return NSLocalizedString( @"NoLikersMessage", @"" );
}

- (UIImage *)noContentImage
{
    return [UIImage imageNamed:@"noLikersIcon"];
}

- (void)refreshWithPageType:(VPageType)pageType completion:(void(^)(BOOL success, NSError *error))completion
{
    [[VObjectManager sharedManager] likersForSequence:self.sequence pageType:pageType
                                         successBlock:^(NSOperation *operation, id result, NSArray *resultObjects)
     {
         completion( YES, nil );
         
     } failBlock:^(NSOperation *operation, NSError *error)
     {
         completion( NO, error );
     }];
}

- (NSArray *)users
{
    return self.sequence.likers.allObjects;
}

@end

File: victorious/victorious/UI/Like/VLikersDataSource.swift
//
//  VLikersDataSource.swift
//  victorious
//
//  Created by Patrick Lynch on 12/21/15.
//  Copyright © 2015 Victorious. All rights reserved.
//

import Foundation

@objc protocol VUsersDataSource: NSObjectProtocol {
    
    func loadUsersWithPageType( pageType: VPageType, completion: (NSError? -> ())?)
    
    func users() -> NSOrderedSet
    
    func noContentTitle() -> String
    
    func noContentMessage() -> String
    
    func noContentImage() -> UIImage
    
    var state: VDataSourceState { get }
    
    var delegate: VPaginatedDataSourceDelegate? { get set }
}

@objc final class VLikersDataSource: PaginatedDataSource, VUsersDataSource {
    
    let sequence: VSequence
    
    init( sequence: VSequence ) {
        self.sequence = sequence
    }
    
    func users() -> NSOrderedSet {
        return self.visibleItems
    }
    
    func loadUsersWithPageType( pageType: VPageType, completion: (NSError? -> ())? = nil ) {
        
        self.loadPage( pageType,
            createOperation: {
                return SequenceLikersOperation(sequenceID: self.sequence.remoteId)
            },
            completion:{ (results, error, cancelled) in
                completion?( error )
            }
        )
    }
    
    // MARK: - VUsersDataSource
    
    func noContentTitle() -> String {
        return  NSLocalizedString( "NoLikersTitle", comment: "" )
    }
    
    func noContentMessage() -> String {
        return  NSLocalizedString( "NoLikersMessage", comment: "" )
    }
    
    func noContentImage() -> UIImage {
        return UIImage(named: "noLikersIcon" )!
    }
}
